### DISCUSSION TITLE
rectangle rotate function

### DISCUSSION THREAD

**Discussion by HeeTop** [wkentaro/labelme#1374] [Ideas / Feature requests]
I would appreciate it if you could create a function to rotate a rectangle.

**Comment by minewilliam** [comment]
I second this!

Having previously worked with oriented bounding boxes, I can pitch in with a feature suggestion.

For legacy reasons, I would opt for a new shape type, call it `rectangle2` . This is because people have built custom parsers/ingestion for the current data format, and adding a field to the current rectangle structure could break old data.

## Rectangle2 definition
There are multiple ways to represent a rectangle 2: (see [YOLO-OBB](https://docs.ultralytics.com/datasets/obb/#yolo-obb-format))
<img width="1920" height="1153" alt="image" src="https://github.com/user-attachments/assets/0ce66091-3ac2-4e95-8f8a-2bae4c459330" />

Out of simplicity, we can use a point-based representation, which is consistent with how shapes are currently stored.
I understand this may be confusing for an end user wishing a different representation. I suggest we could provide some sample Python functions to translate point-based data to a (center, width, height, rotation) representation.

To determine the normal (direction) of the angle, we can also follow the logic used for YOLO-OBB.

## Draw a rectangle2
Because there are many ways of drawing a rotated rectangle, I would personally like to see a 3 point method like below. I find it easier to use than drawing a rectangle from the center. This is also consistent with how other shape types are drawn.
![Screencast_20260106_114132](https://github.com/user-attachments/assets/c5c4f3fc-8629-4fd6-bab7-a37602525f1e)

## Edit a rectangle2
Here, I think it's important that the grab points feel natural:
- Drag a specific edge to translate it perpendicularly.
- Drag a corner to translate two perpendicular edges simultaneously.
- An edge midpoint (rotation) indicator appears when the cursor is near it.
- Drag an edge midpoint to rotate the shape.
- The shape can be moved by dragging the inside of the shape, like any other shape.
![Screencast_20260106_120712](https://github.com/user-attachments/assets/69c70f20-39b3-4752-90ed-7d6f6d8add28)

I hope to reopen this discussion and help speed up the development of this feature.
I am also willing to contribute if the author agrees with the proposition.

**Reply by kancheng** [reply]
@minewilliam 

I’ve also encountered very similar requirements in real-world work settings, especially when dealing with industrial inspection and medical annotation workflows where rotated or oriented rectangles are unavoidable. In those cases, axis-aligned rectangles are often insufficient.

If there is interest from the maintainer side, I’d be happy to collaborate on a dedicated experimental feature branch that implements a minimal but complete and testable version of this functionality. This could allow the author to review a concrete implementation and decide whether the approach fits the project’s long-term direction.

**Comment by kancheng** [comment]
@minewilliam @wkentaro 

I have a quick question regarding the proposal discussed in
https://github.com/wkentaro/labelme/discussions/1517.
https://github.com/wkentaro/labelme/pull/1748

Would it be acceptable to enter the proposed rotated-rectangle (rectangle2) drawing mode via a keyboard modifier while staying in the rectangle tool?
For example, holding a key such as Alt to temporarily switch into this mode, instead of introducing a separate toolbar button or persistent mode switch.

The idea would be:

Default rectangle behavior remains unchanged

Holding a modifier key enables the alternative drawing interaction

Releasing the key returns to the normal rectangle workflow

This would keep the interaction lightweight and avoid breaking existing workflows, while still allowing experimentation with the rectangle2 concept !??????

**Reply by minewilliam** [reply]
I do not believe this is suitable, for a couple of reasons. I believe that rectangle2 needs it's own shape/tool type.

- The data format for a rectangle is two points, while my proposed format for a rotated rectangle is four points. In my opinion, adding more points (from 2 to 4) to the rectangle type would be very confusing and a breaking change for current users.

- It could create a case where the user unknowingly added a tiny angle to their rectangle, basically switching from a 2 point format to a 4 point format.

- We may want to add rotation (mid-)points, to grab onto and apply rotation with. These rotations points would be on every axis-aligned rectangle, and further confuse the user as to which kind of shape/tool they are dealing with.

- To draw a rectangle2, you need 2 anchor points, instead of 1: You need to draw a first line (with points p1 -> p2, which gives the slope) and then a perpendicular line (p2 -> p3 ⟂ p1 -> p2). This difference should be enough to have the user select for a different drawing tool.

FYI. I plan to have a working draft PR for the feature some time next week.

**Reply by kancheng** [reply]
@minewilliam 

Je trouve cela très prometteur!! 

If there’s a pull request, I’m happy to help with testing.

**Reply by wkentaro** [reply]
@minewilliam Thank you for the well-thought out outline. Makes sense to have a new shape_type, and creat-shape-tool.

**Comment by minewilliam** [comment]
@wkentaro 
I believe that this feature may introduce a notable behavior difference with the existing shapes, and maybe make you reconsider restricting annotations to the bounds of the canvas.

Currently, none of the shapes support drawing outside of the canvas (pixmap), which makes sense for some use cases (polygon clipping, for example).
But for this rotated rectangle feature, it is [suitable](https://blog.roboflow.com/labelimg/#:~:text=way%2E-,For,way) to draw outside the canvas, because the object may be partially out of frame. If we restrict a rotated rectangle within the bounds of the canvas, we might as well use polygons, because the proposed shape type cannot be used to label properly.

Here's an example ([image credit](https://home.howstuffworks.com/green-living/uses-old-pill-bottles.htm)) where bottle 1 cannot be annotated because a small part of it lies outside of the canvas. Bottle 2 shows a normal annotation.
<img width="428" height="487" alt="image" src="https://github.com/user-attachments/assets/7804d202-12a6-40eb-ba05-8d50392355a4" />

I am curious to hear your thoughts on drawing outside of the canvas, and if it could apply to other shapes (like rectangles and circles).

**Reply by wkentaro** [reply]
Interesting. Sounds like we should first introduce an option to go beyond image border for any shapes (default can be "auto" or "disabled" to keep the current behavior)

**Reply by minewilliam** [reply]
Created a new discussion for this feature: https://github.com/wkentaro/labelme/discussions/1777

**Comment by minewilliam** [comment]
Found a somewhat relevant discussion: https://github.com/wkentaro/labelme/discussions/1374
The user requested for rotating rectangles, among other shapes.

Included here for posterity.
